### PR TITLE
fix: type completions additonal fixes

### PR DESCRIPTION
**Pull request by dos65**
- fixes completion for term symbol that aren't method
- allow object at type completion pos

 should include a fix for #4224

### CODE REVIEW

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (dos65) [inline mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:70]: @vzmerr I'm sorry that I missed that thing in your initial PR.
In general this was correct but I reworked it a bit at to operate over a rule that we have 3 types of completions.
It's either a type or term or import.
So, having it we can safely filter out symbols: non type at TypePos, non Term at TermPos. For Import it's ok to reference Term and Type.

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (dos65) [inline mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:567]: This was not correct because symbol might be at least also a ValDef.
Also, my suggestion of checking if it's a trait over a `.info` was wrong :facepalm: 
```scala
def foo(param: java.util.List[A]): Unit =
  para@@ // <- here param is not a Method
        // <- also `sym.info.typeSymbol.is(Trait) == true` here because the of param is `java.util.List` 
```

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (vzmerr) [reply mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:70]: Thanks for working on this. May I know the cases where we had the regressions?

File: tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala
Comment (vzmerr) [inline tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala:347]: ArrayDeque takes type parameters. doesn't it? So why is this removed?

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (vzmerr) [reply mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:567]: This method is only excluding class or trait where they are not valid for pos. It does not affect `ValDef`

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (dos65) [reply mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:70]: These two:
https://github.com/scalameta/metals/pull/4234/files#diff-9ab9c5f7b86393976b3619ea86ddef73737390ca43e2a3d8ce27d66aa1f576e7R1534-R1570

File: tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala
Comment (vzmerr) [reply tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala:347]: The current one doesn't compile. It should have type parameters.

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (vzmerr) [reply mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:70]: My current PR on object completion also brings substantial change, based on the existing code. Now, if we use this reworked code, how can we work that out?

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala
Comment (vzmerr) [inline tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:1555]: this one passes on the other PR's branch

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala
Comment (vzmerr) [inline tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:1535]: this one is failing on the other PR's branch. Can we solve it in another way without a total rework, since now on this  branch we have the regression of not adding type parameter square brackets in here.

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala
Comment (vzmerr) [reply tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:1555]: passing on the new PR I created for allowing objects at type pos.

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala
Comment (vzmerr) [reply tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:1535]: perhaps, we can just remove the filter for `isNotClassOrTraitOrTheyAreValidForPos` for now to fix this regression.

Comment (vzmerr) [reply tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:1535]: by simply adding `|| sym.isTerm` to `isNotClassOrTraitOrTheyAreValidForPos` the regression would be solved.

File: tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala
Comment (dos65) [reply tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala:347]: I would turn into an issue after merge

I don't know why it's completed with `[]` in the main branch. 
There is no handler that marks this position by `isTypePosition` for the tree in main  or this case here: 
https://github.com/scalameta/metals/blob/c328f0925e9692b493aba395492ae03fd36da424/mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala#L121-L152
While the tree is: `Ident(ArrayDeque), Typed(Ident(_),Ident(ArrayDeque)), Bind(x,Typed(Ident(_),Ident(ArrayDeque))), CaseDef(Bind(x,Typed(Ident(_),Ident(ArrayDeque)))`

Comment (vzmerr) [reply tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala:347]: Perhaps, this would lead to not adding square brackets also in other positions. Also, what about the other PR with substantial changes that would not be usable after this merge? Also, this is at the last minute, what if it introduces other problems? The regression can be solved by simply adding `|| sym.isTerm` in `isNotClassOrTraitOrTheyAreValidForPos` 
What if the changes of the other PR cannot be reworked on this one?

**Comment by tanishiking** [comment]
Thank you very much @dos65 @vzmerr !
I'm considering maybe we should revert type completions PRs for now, and retry introducing the feature in the next release (0.11.9?).

- https://github.com/scalameta/metals/pull/4174
- https://github.com/scalameta/metals/pull/4174

As @vzmerr mentioned

> Also, this is at the last minute, what if it introduces other problems? The regression can be solved by simply adding || sym.isTerm in isNotClassOrTraitOrTheyAreValidForPos

I'm worried about introducing another problem by this PR (even if adding `|| sym.isTerm` in `isNotClassOrTraitOrTheyAreValidForPos` thing solves this issue) or maybe there's another issue related to type completion. I prefer to skip introducing type completion for 0.11.8 and rework for 0.11.9
What do you think about it? @dos65 @vzmerr FYI @tgodzik @ckipp01

File: tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala
Comment (dos65) [reply tests/cross/src/test/scala/tests/pc/CompletionWorkspaceSuite.scala:347]: yep, `isTerm` fixes this case but we also need to deal with objects.
I've checked that tests doesn't pass in your PR.

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (tgodzik) [reply mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:567]: Looks like it's explained by @dos65  : `sym.info.typeSymbol.is(Trait) == true`

File: tests/cross/src/test/scala/tests/pc/CompletionSuite.scala
Comment (tgodzik) [reply tests/cross/src/test/scala/tests/pc/CompletionSuite.scala:1535]: I think this refactoring is fine and might be more maintainable.

**Review by tgodzik — COMMENTED** [review]
> Thank you very much @dos65 @vzmerr ! I'm considering maybe we should revert type completions PRs for now, and retry introducing the feature in the next release (0.11.9?).
> 
> * [Scala 3 type completion #4174](https://github.com/scalameta/metals/pull/4174)
> * [bugfix: Show package completions #4223](https://github.com/scalameta/metals/pull/4223)
> 
> As @vzmerr mentioned
> 
> > Also, this is at the last minute, what if it introduces other problems? The regression can be solved by simply adding || sym.isTerm in isNotClassOrTraitOrTheyAreValidForPos
> 
> I'm worried about introducing another problem by this PR (even if adding `|| sym.isTerm` in `isNotClassOrTraitOrTheyAreValidForPos` thing solves this issue) or maybe there's another issue related to type completion. I feel it would be safer if we skip introducing type completion for 0.11.8 and rework for 0.11.9 What do you think about it? @dos65 @vzmerr FYI @tgodzik @ckipp01

That would be safest to do probably right now. We could revert the 2 PRs then squash the commits from those two PRs with this one and continue working on this in a new PR that would also handle cases @vzmerr worked on.

@tanishiking could you create the revert and modify the release notes? We would go ahead with the release then to unblock all the waiting PRs.

**Comment by dos65** [comment]
> I'm considering maybe we should revert type completions PRs for now, and retry introducing the feature in the next release (0.11.9?).

@tanishiking Yep, good idea. It seems that we need an additional time to cover all cases (like patterns in case def but I think there should be more) properly and test them before release.

**Comment by tanishiking** [comment]
Thank you, everyone. Let's revert those PRs for now and go on the release.

**Comment by vzmerr** [comment]
> Thank you very much @dos65 @vzmerr ! I'm considering maybe we should revert type completions PRs for now, and retry introducing the feature in the next release (0.11.9?).
> 
> * [Scala 3 type completion #4174](https://github.com/scalameta/metals/pull/4174)
> * [bugfix: Show package completions #4223](https://github.com/scalameta/metals/pull/4223)
> 
> As @vzmerr mentioned
> 
> > Also, this is at the last minute, what if it introduces other problems? The regression can be solved by simply adding || sym.isTerm in isNotClassOrTraitOrTheyAreValidForPos
> 
> I'm worried about introducing another problem by this PR (even if adding `|| sym.isTerm` in `isNotClassOrTraitOrTheyAreValidForPos` thing solves this issue) or maybe there's another issue related to type completion. I feel it would be safer if we skip introducing type completion for 0.11.8 and rework for 0.11.9 What do you think about it? @dos65 @vzmerr FYI @tgodzik @ckipp01

Yes, I think this is the best route. There will be enough time to see if also my new PR can be reworked on this one; and whether this new approach can catch and cover all the cases.

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (vzmerr) [inline mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:165]: why is this TermPos? It does not differentiate between `CursorPositionCondition(false, false, true, hasNoSquareBracket)` and this one

File: mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala
Comment (vzmerr) [inline mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:146]: in here, it was intentional that the `if` guard was no put in the case. It creates a different logic when you add the guard to case. The `else` is different with the last `_` case that we have.

Comment (vzmerr) [reply mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala:146]: I think the tweaking of logic in here, is the cause of not inserting square bracket in the test that is not passing.
`pprints` would be showing that.

**Review by tanishiking — APPROVED** [review]
What is the status of this PR? I took a look and its LGTM 💯 

We talked with @dos65 in person, and we may want to re-work on type of completion stuff, but we can work on it in another PR (as this PR already fixed some regressions 👍 ).

### what kind of re-work we talked about?
Currently, type completion filters out "non-compilable completions" such as `Foo` (where `Foo` should have type parameters list `Foo[Int]`).

However, it would be nice if we provide both `Foo` (with lower priority) and `Foo[$0]` (with higher priority) instead of filtering out `Foo` (without brackets). Providing both of them makes Metals easier to introduce unexpected regressions.

---

(side note) My thoughts on completions in IDE

I believe completing just `Foo` is still fine as I don't expect completions to be compilable out of the box. Completions provide some help to developers, they don't need to be "semantically correct".

For example,

```scala
def foo(x: Int) = ???
val aaa: Int = 1
val abb: String = "1"

foo(a@@) // should we filter out `abb` from completion items
// because `abb`'s return type is String
// whereas `foo` should receive Int parameters?

// I don't think so, maybe users want to type `abb.toInt` after completing `abb`
```

**Pull request merged**
